### PR TITLE
ipatests: Add AD trust SELinux user map tests (part 1)

### PR DESCRIPTION
**Pull request by pranav210798**
Add TestTrustFunctionalSelinuxUsermap with trust setup, groups setup, and:
- 001: Apply a SELinux user map to one client only; mapped users get the map context on that host and the default context elsewhere.
- 002: Apply a SELinux user map to the IPA master only; mapped users get the map context when logging in to the master and the default elsewhere.
- 003: Tie a SELinux user map to an HBAC rule on one client; access and SELinux context follow HBAC plus the map on that host.

Related : https://pagure.io/freeipa/issue/9845

## Summary by Sourcery

Add integration test coverage for SELinux user maps in IPA–AD trust setups, including host-specific, master-specific, and HBAC-linked mappings for root and subdomain AD users.

Enhancements:
- Extend integration task helpers with SELinux user map management wrappers and make HBAC rule deletion optionally non-fatal to support test cleanup.

Tests:
- Introduce TestTrustFunctionalSelinuxUsermap suite exercising SELinux user map behavior for AD trust users across two IPA clients and the master, including maps bound to specific hosts and HBAC rules.
- Add SSH-based verification helpers to assert SELinux contexts returned for trusted users when authenticating with Kerberos credentials and passwords.

### CODE REVIEW

File: ipatests/test_integration/test_trust_functional.py [hunk 1]
```diff
@@ -2579,3 +2672,370 @@ def test_group_memberships_resolve_ad_users(self):
...
+                ['systemctl', 'enable', '--now', 'oddjobd']
+            )
+
+    def _selinux_hosts(self):
+        return (self.master, self.client1, self.client2)
+
+    def _selinux_cache_reset_all(self):
+        for host in self._selinux_hosts():
+            tasks.clear_sssd_cache(host)
+            try:
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: No need to add try:, except block for `tasks.wait_for_sssd_domain_status_online(host)`

File: ipatests/test_integration/test_trust_functional.py [hunk 2]
```diff
@@ -2579,3 +2672,370 @@ def test_group_memberships_resolve_ad_users(self):
...
+                tasks.wait_for_sssd_domain_status_online(host)
+            except Exception:
+                pass
+        time.sleep(10)
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Remove sleep.

File: ipatests/test_integration/test_trust_functional.py [hunk 3]
```diff
@@ -2579,3 +2672,370 @@ def test_group_memberships_resolve_ad_users(self):
...
+
+    def _cleanup_selinux_groups(self, case):
+        tasks.kinit_admin(self.master)
+        for grp in (case['group1_ext'], case['group1'],
+                    case['group2_ext'], case['group2']):
+            self.master.run_command(
+                ['ipa', 'group-del', grp], raiseonerr=False
+            )
+
+    def test_selinuxusermap_trust_setup(self):
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: No need to add new test. This can be moved to install classmethod

File: ipatests/test_integration/test_trust_functional.py [hunk 4]
```diff
@@ -2579,3 +2672,370 @@ def test_group_memberships_resolve_ad_users(self):
...
+        tasks.establish_trust_with_ad(
+            self.master, self.ad_domain,
+            extra_args=['--range-type', 'ipa-ad-trust'],
+        )
+        tasks.kinit_admin(self.master)
+
+    def test_selinuxusermap_groups_setup(self):
+        """Add AD trust users to POSIX groups (per root and subdomain)."""
+        for case in self._selinux_domain_cases():
+            self._setup_selinux_groups(case)
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Same here as well. This can be moved to install classmethod.

File: ipatests/test_integration/test_trust_functional.py [hunk 5]
```diff
@@ -2579,3 +2672,370 @@ def test_group_memberships_resolve_ad_users(self):
...
+
+    def test_selinuxusermap_001_map_on_particular_host(self):
+        """Map AD users to SELinux users on a particular host (001)."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Add more detailed description in docstring. This test is performing multiple checks.

File: ipatests/test_integration/test_trust_functional.py [hunk 6]
```diff
@@ -2579,3 +2672,370 @@ def test_group_memberships_resolve_ad_users(self):
...
+                    self.client2, user1, self.ad_user_password,
+                    self.client2, SELINUX_DEFAULT_VERIF)
+            finally:
+                tasks.kinit_admin(self.master)
+                tasks.selinuxusermap_del(
+                    self.master, 'selinuxusermaprule1', raiseonerr=False,
+                )
+
+    def test_selinuxusermap_002_map_on_master(self):
+        """Map AD users to SELinux users on the IPA master (002)."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Same here as well. Add more detailed description in docstring. This test is performing multiple checks.

File: ipatests/test_integration/test_trust_functional.py [hunk 7]
```diff
@@ -2579,3 +2672,370 @@ def test_group_memberships_resolve_ad_users(self):
...
+                    self.client2, user2, self.ad_user_password,
+                    self.master, SELINUX_T1_VERIF)
+            finally:
+                tasks.kinit_admin(self.master)
+                tasks.selinuxusermap_del(
+                    self.master, 'selinuxusermaprule2', raiseonerr=False,
+                )
+
+    def test_selinuxusermap_003_hbac_linked_map(self):
+        """HBAC-linked SELinux user map on a specific client (003)."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Same here. Add more detailed description in docstring. This test is performing multiple checks.

**Review by amore17 — COMMENTED** [review]
Hi @pranav210798 
please find inline comments

**Comment by pranav210798** [comment]
@amore17 The changes are done.

File: ipatests/pytest_ipa/integration/tasks.py [hunk 8]
```diff
@@ -2403,7 +2403,7 @@ def hbacrule_remove_service(host, rulename, services=None,
     return host.run_command(cmd, raiseonerr=raiseonerr)
 
 
-def hbacrule_del(host, rulename, extra_args=()):
+def hbacrule_del(host, rulename, extra_args=(), raiseonerr=True):
```
Comment (amore17) [inline ipatests/pytest_ipa/integration/tasks.py:2406]: In temp commit also add job for test_hbac_functional.py to verify these changes are not breaking any other tests.

File: ipatests/pytest_ipa/integration/tasks.py [hunk 9]
```diff
@@ -2421,6 +2421,92 @@ def hbacrule_disable(host, rulename, raiseonerr=True):
...
+
+
+def selinuxusermap_add_host(host, mapname, hosts=None,
+                            raiseonerr=True, extra_args=()):
+    cmd = ["ipa", "selinuxusermap-add-host", mapname]
+    if hosts:
+        cmd.append("--hosts")
+        cmd.append(hosts)
+    cmd.extend(extra_args)
+    return host.run_command(cmd, raiseonerr=raiseonerr)
```
Comment (amore17) [inline ipatests/pytest_ipa/integration/tasks.py]: This can be improved : 
`
def selinuxusermap_add_host(
    host,
    mapname,
    hosts=None,
    raiseonerr=True,
    extra_args=None,
):
    cmd = ["ipa", "selinuxusermap-add-host", mapname]

    if hosts:
        cmd.extend(["--hosts", hosts])

    if extra_args:
        cmd.extend(extra_args)

    return host.run_command(cmd, raiseonerr=raiseonerr)
`

File: ipatests/pytest_ipa/integration/tasks.py [hunk 10]
```diff
@@ -2421,6 +2421,92 @@ def hbacrule_disable(host, rulename, raiseonerr=True):
...
+
+
+def selinuxusermap_remove_host(host, mapname, hosts=None,
+                               raiseonerr=True, extra_args=()):
+    cmd = ["ipa", "selinuxusermap-remove-host", mapname]
+    if hosts:
+        cmd.append("--hosts")
+        cmd.append(hosts)
```
Comment (amore17) [inline ipatests/pytest_ipa/integration/tasks.py]: This can be improved : 
`
    if hosts:
        cmd.extend(["--hosts", hosts])

`

File: ipatests/test_integration/test_trust_functional.py [hunk 11]
```diff
@@ -160,6 +160,99 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+def _ssh_id_z_over_gssapi(host, login, target_hostname, retries=11):
+    """SSH to *target_hostname* as *login* and return ``id -Z`` output."""
+    last = None
+    stdout = None
+    for _attempt in range(retries):
+        result = host.run_command(
+            [
+                'ssh', '-o', 'StrictHostKeyChecking=no', '-K',
+                '-o', 'PubkeyAuthentication=no',
+                '-l', login, target_hostname, 'sleep 10; id -Z',
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Why sleep 10 is required ? If we are running it with retries then this sleep can be removed.

File: ipatests/test_integration/test_trust_functional.py [hunk 12]
```diff
@@ -160,6 +160,99 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+            ],
+            raiseonerr=False,
+        )
+        last = result
+        if result.returncode == 0:
+            stdout = result.stdout_text
+            break
+        time.sleep(20)
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Remove sleep.

**Review by amore17 — COMMENTED** [review]
@pranav210798 
Also general comment. Please remove unnecessary sleep.

File: ipatests/test_integration/test_trust_functional.py [hunk 13]
```diff
@@ -160,6 +160,99 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+SELINUX_T2 = 'user_u:s0'
+SELINUX_T2_VERIF = r'user_u:.*s0'
+SELINUX_T3 = 'xguest_u:s0'
+SELINUX_T3_VERIF = r'xguest_u:.*s0'
+SELINUX_T4 = 'guest_u:s0'
+SELINUX_T4_VERIF = r'guest_u:.*s0'
+
+
+def _ssh_id_z_over_gssapi(host, login, target_hostname, retries=11):
+    """SSH to *target_hostname* as *login* and return ``id -Z`` output."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: use Sphinx-style [docstrings](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

File: ipatests/test_integration/test_trust_functional.py [hunk 14]
```diff
@@ -160,6 +160,99 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+        f'SSH with GSSAPI from {host.hostname} as {login} to '
+        f'{target_hostname} failed on attempt {_attempt + 1}/{retries}: '
+        f'{output}'
+    )
+    return ''
+
+
+def verify_ssh_selinuxuser_success_with_krbcred(
+        host, login, target_host, selinux_pattern):
+    """Assert GSSAPI SSH yields the expected SELinux user context."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: use Sphinx-style [docstrings](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

File: ipatests/test_integration/test_trust_functional.py [hunk 15]
```diff
@@ -160,6 +160,99 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+    output = _ssh_id_z_over_gssapi(host, login, target_host.hostname)
+    assert re.search(selinux_pattern, output), (
+        f'Expected SELinux context matching {selinux_pattern!r} for {login} '
+        f'on {target_host.hostname}, got: {output!r}'
+    )
+
+
+def verify_ssh_selinuxuser_failure_with_krbcred(
+        host, login, target_host, selinux_pattern):
+    """Assert GSSAPI SSH does not yield *selinux_pattern*."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: use Sphinx-style [docstrings](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

File: ipatests/test_integration/test_trust_functional.py [hunk 16]
```diff
@@ -160,6 +160,99 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+    output = _ssh_id_z_over_gssapi(host, login, target_host.hostname)
+    assert not re.search(selinux_pattern, output), (
+        f'Unexpected SELinux context matching {selinux_pattern!r} for {login} '
+        f'on {target_host.hostname}, got: {output!r}'
+    )
+
+
+def verify_ssh_auth_success_selinuxuser(
+        ssh_host, login, password, target_host, selinux_pattern):
+    """Assert password SSH yields the expected SELinux user context."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: use Sphinx-style [docstrings](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

File: ipatests/test_integration/test_trust_functional.py [hunk 17]
```diff
@@ -160,6 +160,99 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+    output = result.stdout_text
+    assert re.search(selinux_pattern, output), (
+        f'Expected SELinux context matching {selinux_pattern!r} for {login} '
+        f'on {target_host.hostname}, got: {output!r}'
+    )
+
+
+def verify_ssh_auth_failure_selinuxuser(
+        ssh_host, login, password, target_host, selinux_pattern):
+    """Assert password SSH does not yield *selinux_pattern*."""
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: use Sphinx-style [docstrings](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

**Comment by rjeffman** [comment]
@pranav210798 as we discussed, here's the automated review, without me reviewing it, for **test purposes only**. Use at your own risk. ;-)

---
  I've completed a comprehensive multi-agent review of PR #8409 from freeipa/freeipa. The review was conducted by 5 specialized agents working in parallel:                                    
   
  1. code-reviewer - Code quality and maintainability                                                                                                                                          
  2. security-reviewer - Security and IAM aspects
  3. adversarial-qe - Adversarial testing for bugs and edge cases
  4. test-design-expert - Test coverage and design quality
  5. iam-engineer - Authentication, authorization, and trust boundary enforcement

  Summary

  Total Findings: 28 (3 Critical, 6 High, 13 Medium, 6 Low)

  Critical Issues (Must Fix Before Merge)

  1. hbacrule_del bug - The raiseonerr parameter is accepted but never forwarded to host.run_command(), breaking cleanup in test_003
  2. No retry delay - _ssh_id_z_over_gssapi retries 11 times with zero sleep between attempts, causing flaky tests
  3. Missing function - tasks.wait_for_sssd_domain_status_online() is called but doesn't exist (AI hallucination) - every test will crash

  High-Priority Issues

  - .freeipa-pr-ci.yaml must be reverted to gating.yaml (currently points to temp config)
  - Test_003 doesn't disable allow_all HBAC rule, so it only tests SELinux mapping, not actual HBAC access denial
  - Verification functions only check negative assertions, not positive (won't catch broken systems)
  - Missing assertions for user2 default context in test_002

  Full Report

  The complete review has been saved to:
  /home/rjeffman/Developer/agentic_idm/code_review/staging/reviews/2026-05-21-pr-8409-selinux-usermap-review.md

  The report includes:
  - Detailed findings with file locations, line numbers, and code snippets
  - Impact analysis and fix recommendations for each issue
  - Test coverage gap analysis
  - Prioritized action plan
  - Positive observations about what's done well

  Recommendation: Approve after fixing the 3 critical issues and reverting the CI config. The other high-priority items should be addressed for production-quality testing.

File: ipatests/pytest_ipa/integration/tasks.py
Comment (rjeffman) [inline ipatests/pytest_ipa/integration/tasks.py]: Why not use raiseonerr on this call?

File: ipatests/test_integration/test_trust_functional.py [hunk 18]
```diff
@@ -160,6 +160,136 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+                'ssh', '-o', 'StrictHostKeyChecking=no', '-K',
+                '-o', 'PubkeyAuthentication=no',
+                '-l', login, target_hostname, 'id -Z',
+            ],
+            raiseonerr=False,
+        )
+        last = result
+        if result.returncode == 0:
+            stdout = result.stdout_text
+            break
```
Comment (rjeffman) [inline ipatests/test_integration/test_trust_functional.py]: There should be a (random?) wait here to give the system some chance to recover, rather then trying again directly.

File: ipatests/test_integration/test_trust_functional.py [hunk 19]
```diff
@@ -2579,3 +2709,379 @@ def test_group_memberships_resolve_ad_users(self):
...
+            ssh_with_password(
+                self.master, user1, self.master.hostname,
+                self.ad_user_password,
+            )
+            ssh_with_password(
+                self.master, user2, self.master.hostname,
+                self.ad_user_password,
+            )
+            tasks.kinit_admin(self.master)
+            try:
```
Comment (rjeffman) [inline ipatests/test_integration/test_trust_functional.py]: What if the previous ssh_with_password failed completely? Should this step still be executed?

Comment (pranav210798) [reply ipatests/test_integration/test_trust_functional.py]: Yes this should also fail completely.

File: ipatests/pytest_ipa/integration/tasks.py
Comment (rjeffman) [reply ipatests/pytest_ipa/integration/tasks.py]: yes, but if one say `raiseonerr=False` it will not be passed to run_command in this case, going against the function contract.

File: ipatests/test_integration/test_trust_functional.py (hunk 19 above)
Comment (rjeffman) [reply ipatests/test_integration/test_trust_functional.py]: Wouldn't it be better to fail early?

File: ipatests/pytest_ipa/integration/tasks.py
Comment (pranav210798) [reply ipatests/pytest_ipa/integration/tasks.py]: Yeah, i missed that completely.

File: ipatests/test_integration/test_trust_functional.py (hunk 19 above)
Comment (pranav210798) [reply ipatests/test_integration/test_trust_functional.py]: I pushed the ssh_with_password into try block, if it fails the whole test fails and it will go to cleanup.

File: ipatests/pytest_ipa/integration/tasks.py [hunk 20]
```diff
@@ -2421,6 +2421,102 @@ def hbacrule_disable(host, rulename, raiseonerr=True):
     return host.run_command(cmd, raiseonerr=raiseonerr)
 
 
+def selinuxusermap_add(host, mapname, extra_args=()):
+    cmd = ["ipa", "selinuxusermap-add", mapname]
+    cmd.extend(extra_args)
+    return host.run_command(cmd)
```
Comment (duzda) [inline ipatests/pytest_ipa/integration/tasks.py:2427]: You can do `cmd = ["ipa", "selinuxusermap-add", mapname, *extra_args]`, also as a note, you define the type as tuple here, but later assign array, this does work, but is a different type. I don't propose to change that, as it's very common pattern in the codebase, but it's strange to see it in use.

File: ipatests/pytest_ipa/integration/tasks.py [hunk 21]
```diff
@@ -2421,6 +2421,102 @@ def hbacrule_disable(host, rulename, raiseonerr=True):
...
+    if groups:
+        cmd.append(f"--groups={groups}")
+    cmd.extend(extra_args)
+    return host.run_command(cmd, raiseonerr=raiseonerr)
+
+
+def selinuxusermap_add_host(
+        host,
+        mapname,
+        hosts=None,
```
Comment (duzda) [inline ipatests/pytest_ipa/integration/tasks.py:2456]: Rename to host ~~s~~, or rename function to add_host**s**

File: ipatests/test_integration/test_trust_functional.py [hunk 22]
```diff
@@ -16,7 +16,17 @@
 from ipatests.util import xfail_context
 
 PLUGIN_CONF = "/var/lib/sss/pubconf/krb5.include.d/localauth_plugin"
-
+# SELinux user contexts used by ipa-trust-functional selinuxusermap tests
+SELINUX_DEFAULT = 'unconfined_u:s0-s0:c0.c1023'
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: Unused

File: ipatests/test_integration/test_trust_functional.py [hunk 23]
```diff
@@ -16,7 +16,17 @@
...
+SELINUX_DEFAULT_VERIF = r'unconfined_u:.*s0-s0:c0.c1023'
+SELINUX_T1 = 'staff_u:s0-s0:c0.c1023'
+SELINUX_T1_VERIF = r'staff_u:.*s0-s0:c0.c1023'
+SELINUX_T2 = 'user_u:s0'
+SELINUX_T2_VERIF = r'user_u:.*s0'
+SELINUX_T3 = 'xguest_u:s0'
+SELINUX_T3_VERIF = r'xguest_u:.*s0'
+SELINUX_T4 = 'guest_u:s0'
+SELINUX_T4_VERIF = r'guest_u:.*s0'
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: Unused

I understand that this is part 1, consider the `_` prefix, so lsp can tell you about unused variables in case of leftovers.

File: ipatests/test_integration/test_trust_functional.py [hunk 24]
```diff
@@ -160,6 +170,124 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
     )
 
 
+def _ssh_id_z_over_gssapi(host, login, target_hostname, retries=11):
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: 11 is a strange choice, 10 is nicer

File: ipatests/test_integration/test_trust_functional.py [hunk 25]
```diff
@@ -160,6 +170,124 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+
+    :param host: Host from which SSH is run (must already have a Kerberos
+        ticket for ``login``).
+    :param login: SSH login name (``-l`` argument).
+    :param target_hostname: Hostname to connect to.
+    :param retries: Number of attempts before failing (default 11).
+    :returns: ``stdout`` from a successful ``id -Z`` run.
+    :raises pytest.fail: If SSH does not succeed within ``retries`` attempts.
+    """
+    last = None
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: You can remove this  variable all together, Python doesn't remove variables out of what other languages consider "block" scope.

The following is perfectly valid Python code:

```python
for i in range(10):
  if True:
    x = i + i

print(x)
```

Function scope is the smallest scope you get in Python.

File: ipatests/test_integration/test_trust_functional.py [hunk 26]
```diff
@@ -160,6 +170,124 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+        return stdout
+    output = ''
+    if last is not None:
+        output = f'{last.stdout_text}{last.stderr_text}'
+    pytest.fail(
+        f'SSH with GSSAPI from {host.hostname} as {login} to '
+        f'{target_hostname} failed on attempt {_attempt + 1}/{retries}: '
+        f'{output}'
+    )
+    return ''
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: This is unreachable, the line above throws.

File: ipatests/test_integration/test_trust_functional.py [hunk 27]
```diff
@@ -160,6 +170,124 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+            [
+                'ssh', '-o', 'StrictHostKeyChecking=no', '-K',
+                '-o', 'PubkeyAuthentication=no',
+                '-l', login, target_hostname, 'id -Z',
+            ],
+            raiseonerr=False,
+        )
+        last = result
+        if result.returncode == 0:
+            stdout = result.stdout_text
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: You can instead do `return result.stdout_text` and remove the `stdout` variable.

File: ipatests/test_integration/test_trust_functional.py [hunk 28]
```diff
@@ -160,6 +170,124 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+    :param selinux_pattern: Regex matched against ``id -Z`` output.
+    """
+    output = _ssh_id_z_over_gssapi(host, login, target_host.hostname)
+    assert re.search(selinux_pattern, output), (
+        f'Expected SELinux context matching {selinux_pattern!r} for {login} '
+        f'on {target_host.hostname}, got: {output!r}'
+    )
+
+
+def verify_ssh_selinuxuser_failure_with_krbcred(
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: It would make more sense to simply return boolean and let the user do the assert themselves

File: ipatests/test_integration/test_trust_functional.py [hunk 29]
```diff
@@ -160,6 +170,124 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+    result = ssh_with_password(
+        ssh_host, login, target_host.hostname, password,
+        remote_cmd='id -Z',
+        expect_success=True,
+    )
+    output = result.stdout_text
+    assert not re.search(selinux_pattern, output), (
+        f'Unexpected SELinux context matching {selinux_pattern!r} for {login} '
+        f'on {target_host.hostname}, got: {output!r}'
+    )
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: The same with these two identical functions

File: ipatests/test_integration/test_trust_functional.py [hunk 30]
```diff
@@ -160,6 +160,93 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+        f'Expected SELinux context matching {selinux_pattern!r} for {login} '
+        f'on {target_host.hostname}, got: {output!r}'
+    )
+
+
+def verify_ssh_selinuxuser_failure_with_krbcred(
+        host, login, target_host, selinux_pattern):
+    """Assert GSSAPI SSH does not yield *selinux_pattern*."""
+    output = _ssh_id_z_over_gssapi(host, login, target_host.hostname)
+    assert not re.search(selinux_pattern, output), (
```
[the remark this answers is not in the document]
Comment (duzda) [reply ipatests/test_integration/test_trust_functional.py]: See the comment here?

File: ipatests/test_integration/test_trust_functional.py [hunk 31]
```diff
@@ -2579,3 +2707,383 @@ def test_group_memberships_resolve_ad_users(self):
...
+    topology = 'line'
+    num_clients = 2
+    num_ad_treedomains = 0
+    ad_user_password = 'Secret123'
+
+    @classmethod
+    def install(cls, mh):
+        super().install(mh)
+        cls.client1 = cls.clients[0]
+        cls.client2 = cls.clients[1]
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: This is completely unnecessary and unsure whether this "improves" readability, I say not really.

File: ipatests/test_integration/test_trust_functional.py [hunk 32]
```diff
@@ -2579,3 +2707,383 @@ def test_group_memberships_resolve_ad_users(self):
...
+        tasks.kinit_admin(cls.master)
+        for host in (cls.master, cls.client1, cls.client2):
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: (cls.master, *cls.clients)

File: ipatests/test_integration/test_trust_functional.py [hunk 33]
```diff
@@ -2579,3 +2707,383 @@ def test_group_memberships_resolve_ad_users(self):
...
+            cls.master, cls.ad_domain,
+            extra_args=['--range-type', 'ipa-ad-trust'],
+        )
+        tasks.kinit_admin(cls.master)
+        for case in cls._selinux_domain_cases():
+            cls._setup_selinux_groups(case)
+
+    @classmethod
+    def _selinux_cache_reset_all(cls):
+        for host in (cls.master, cls.client1, cls.client2):
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: Similarly here

File: ipatests/test_integration/test_trust_functional.py [hunk 34]
```diff
@@ -2579,3 +2707,383 @@ def test_group_memberships_resolve_ad_users(self):
...
+        base = self._ad_user_base(user_fqdn)
+        return self._ad_principal(base, domain, realm=True)
+
+    def _kinit_trust_user(self, host, user_fqdn, domain):
+        upn = self._user_upn(user_fqdn, domain)
+        tasks.kdestroy_all(host)
+        tasks.kinit_as_user(host, upn, self.ad_user_password)
+
+    @classmethod
+    def _setup_selinux_groups(cls, case):
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py:2753]: It would be great idea to add a comment explaning what `case` is, and that it can be obtained through `_selinux_domain_cases`

File: ipatests/test_integration/test_trust_functional.py [hunk 35]
```diff
@@ -2579,3 +2707,383 @@ def test_group_memberships_resolve_ad_users(self):
...
+                )
+                tasks.selinuxusermap_add_host(
+                    self.master, 'selinuxusermaprule2',
+                    hosts=self.master.hostname,
+                )
+                self._selinux_cache_reset_all()
+
+                user1 = case['user1']
+                user2 = case['user2']
+                domain = case['domain']
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py:2908]: You can abstract a lot of the stuff away, it's repeated throughout the test cases, but I'm fine with copy pasting here so far.

Comment (pranav210798) [reply ipatests/test_integration/test_trust_functional.py:2908]: I will keep it simple for each testcase, rather than making the tests to look complex.

File: ipatests/pytest_ipa/integration/tasks.py (hunk 20 above)
Comment (pranav210798) [reply ipatests/pytest_ipa/integration/tasks.py:2427]: Yeah we are already implementing it in the same way so following the same pattern.

File: ipatests/test_integration/test_trust_functional.py [hunk 36]
```diff
@@ -160,6 +164,100 @@ def passwd_change_with_retry(host, user_fqdn, current_password,
...
+    if expect_match:
+        assert matched, (
+            f'Expected SELinux context matching {selinux_pattern!r} for '
+            f'{login} on {target_hostname}, got: {output!r}'
+        )
+    else:
+        assert not matched, (
+            f'Unexpected SELinux context matching {selinux_pattern!r} for '
+            f'{login} on {target_hostname}, got: {output!r}'
+        )
```
Comment (duzda) [inline ipatests/test_integration/test_trust_functional.py]: You can simplify this even more:

```python
assert matched == expect_match, (
    f'Unexpected SELinux context matching {selinux_pattern!r} for '
    f'{login} on {target_hostname}, got: {output!r}'
)
```

as you're basically doing xor with the if here.

(oh wow, some days I'm really clever!!)

**Review by duzda — APPROVED** [review]
I won't nudge about this simplification, as someone in the future might get stuck on this for a while.

**Comment by duzda** [comment]
Removed 4.12 as 4.12 is unused

**Comment by pranav210798** [comment]
Green Temp Commit: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/373d45a2-583f-11f1-a9c7-fa163e863465

**Comment by duzda** [comment]
@pranav210798  please remove temp commit

**Comment by pranav210798** [comment]
@duzda Removed the temp commit.

**Comment by amore17** [comment]
@pranav210798 Could you please also update prci definitions. These new changes requires new topology adroot_adchild_master_2client compare to other classes.

**Comment by pranav210798** [comment]
@amore17 Can you please review the updated PRCI definitions?
.

**Comment by amore17** [comment]
> @amore17 Can you please review the updated PRCI definitions? .

@pranav210798 It would be better if prci changes in new commit.

**Comment by pranav210798** [comment]
@amore17 Done

**Comment by menonsudhir** [comment]
master:

* e4fa0477bb7d7f91897dc53a72c68bd1c78a86d2 ipatests: Add AD trust SELinux user map tests (part 1)
* 4c98b601c58d041a07629517546327b7d43eaaa0 PRCI definations updated for trust_selinux_usermap.

**Comment by menonsudhir** [comment]
Manual backport to ipa-4-13 needs to be done. @pranav210798 can you take care of the backport.

**Comment by pranav210798** [comment]
> Manual backport to ipa-4-13 needs to be done. @pranav210798 can you take care of the backport.

@menonsudhir Sure will do the manual backports.
